### PR TITLE
Trim entries for which TP + FP = 0

### LINKED ISSUE #444: There is a steep drop at the beginning of PR curves.
... at which both precision and recall are very close to 0.

csq8ynwugjk 1

Lets see if we can fix this within the PR curve summary op.

### PR DESCRIPTION
**Pull request by chihuahua**
We trim entries for which TP + FP = 0 at the tail of data in the backend logic of
the PR curves plugin.

For those values, precision is undefined. Fixes #444. Indeed, the leftmost points 
of PR curves now adhere to TP + FP > 0.

rcvhnhvhzwr

### CODE REVIEW

File: tensorboard/plugins/pr_curve/pr_curves_plugin.py [hunk 1]
```diff
@@ -214,29 +214,42 @@ def is_active(self):
...
 
   def _process_tensor_event(self, event, thresholds):
-    """Converts a TensorEvent into an dict that encapsulates information on it.
+    """Converts a TensorEvent into a dict that encapsulates information on it.
 
     Args:
       event: The TensorEvent to convert.
       thresholds: An array of floats that ranges from 0 to 1 (in that
-        direction).
+        direction). This list of thresholds is not culled. We span the whole
```
Comment (wchargin) [inline tensorboard/plugins/pr_curve/pr_curves_plugin.py]: What does "not culled" mean?

Is "We span the whole range from 0 to 1" not redundant with "An array of floats that ranges from 0 to 1"? If not, please clarify what it actually means.

File: tensorboard/plugins/pr_curve/pr_curves_plugin.py [hunk 2]
```diff
@@ -214,29 +214,42 @@ def is_active(self):
...
       A JSON-able dictionary of PR curve data for 1 step.
     """
     data_array = tf.make_ndarray(event.tensor_proto)
+
+    # Remove entries for which TP + FP = 0 (precision is undefined).
+    true_positives = [int(v) for v in data_array[metadata.TRUE_POSITIVES_INDEX]]
+    false_positives = [
+        int(v) for v in data_array[metadata.FALSE_POSITIVES_INDEX]]
+    positives = tuple(sum(t) for t in zip(true_positives, false_positives))
+    for end_index in range(len(positives) - 1, -1, -1):
```
Comment (wchargin) [inline tensorboard/plugins/pr_curve/pr_curves_plugin.py]: This is alternately written as
```python
end_index_inclusive = len(positives) - 1
while end_index_inclusive > 0 and positives[end_index_inclusive] == 0:
  end_index_inclusive -= 1
end_index = end_index_inclusive - 1
```
which is a bit more readable to me. What do you think?

File: tensorboard/plugins/pr_curve/pr_curves_plugin.py [hunk 3]
```diff
@@ -214,29 +214,42 @@ def is_active(self):
...
   def _process_tensor_event(self, event, thresholds):
-    """Converts a TensorEvent into an dict that encapsulates information on it.
+    """Converts a TensorEvent into a dict that encapsulates information on it.
 
     Args:
       event: The TensorEvent to convert.
       thresholds: An array of floats that ranges from 0 to 1 (in that
-        direction).
+        direction). This list of thresholds is not culled. We span the whole
+        range from 0 to 1. This method filters the thresholds based on the
```
Comment (wchargin) [inline tensorboard/plugins/pr_curve/pr_curves_plugin.py]: After reading this description, I was surprised at what the code actually does. You aren't actually "removing values with TP + FP = 0"; you're only trimming such values from the tail of the data. Do you think that this merits clarification?

File: tensorboard/plugins/pr_curve/pr_curves_plugin.py [hunk 4]
```diff
@@ -214,29 +214,42 @@ def is_active(self):
...
     Returns:
       A JSON-able dictionary of PR curve data for 1 step.
     """
     data_array = tf.make_ndarray(event.tensor_proto)
+
+    # Remove entries for which TP + FP = 0 (precision is undefined).
+    true_positives = [int(v) for v in data_array[metadata.TRUE_POSITIVES_INDEX]]
+    false_positives = [
+        int(v) for v in data_array[metadata.FALSE_POSITIVES_INDEX]]
+    positives = tuple(sum(t) for t in zip(true_positives, false_positives))
```
Comment (wchargin) [inline tensorboard/plugins/pr_curve/pr_curves_plugin.py]: Perhaps more readably:
```py
positives = (
    data_array[[metadata.TRUE_POSITIVES_INDEX, metadata.FALSE_POSITIVES_INDEX], :]
    .astype(int).sum(axis=0).tolist())`
```
(The `.tolist()` isn't really necessary; I just included it to match your current datatypes.)

File: tensorboard/plugins/pr_curve/pr_curves_plugin.py (hunk 1 above)
Comment (chihuahua) [reply tensorboard/plugins/pr_curve/pr_curves_plugin.py]: I think this is more concise and clearer.

File: tensorboard/plugins/pr_curve/pr_curves_plugin.py (hunk 3 above)
Comment (chihuahua) [reply tensorboard/plugins/pr_curve/pr_curves_plugin.py]: That is such a clearer way of stating that.

Comment (chihuahua) [reply tensorboard/plugins/pr_curve/pr_curves_plugin.py]: Also updated the PR description.

File: tensorboard/plugins/pr_curve/pr_curves_plugin.py (hunk 4 above)
Comment (chihuahua) [reply tensorboard/plugins/pr_curve/pr_curves_plugin.py]: Sounds great! I also like how the operations (summation, casting) stay within data array land.

File: tensorboard/plugins/pr_curve/pr_curves_plugin.py (hunk 2 above)
Comment (chihuahua) [reply tensorboard/plugins/pr_curve/pr_curves_plugin.py]: Yeah, lets do that and not abuse the for loop.

**Pull request merged**